compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,511 @@
+    //   3) There are no dmas waiting for this dma to produce data.
+    //
+    // The last two conditions are partially checked through comparing source
+    // and target memory spaces (Global < L2 < L1):
+    //   1) In the forward sweep, only hoist dmas for which
+    //   (source.memory < target.memory), i.e. moving data to AIE cores.
+    //   2) In the backward sweep, only hoist dmas for which
+    //   (source.memory > target.memory), i.e. moving data away from AIE cores.
+    //
+    // These last checks in theory limit the hoisting detection capability, but
+    // should be valid.
+    forOp.walk<WalkOrder::PostOrder, Iterator>([&](AMDAIE::DmaCpyNdOp dmaOp) {
+      if (llvm::any_of(dmaOp->getOpOperands(), [&](OpOperand &operand) {
+            return dependsOnLoop(operand);
+          })) {
+        addDependencies(dmaOp);
+        return WalkResult::advance();
+      }
+
+      uint64_t sourceMemspace =
+          dmaOp.getSourceObjectFifo().getMemorySpaceAsUInt();
+      uint64_t targetMemspace =
+          dmaOp.getTargetObjectFifo().getMemorySpaceAsUInt();
+      if (std::is_same<Iterator, ForwardIterator>::value &&
+          !dependencies.contains(dmaOp.getSourceObjectFifo()) &&
+          sourceMemspace < targetMemspace) {
+        rewriter.moveOpBefore(dmaOp, forOp);
+        hoistHappened = true;
+        return WalkResult::advance();
+      } else if (std::is_same<Iterator, ReverseIterator>::value &&
+                 !dependencies.contains(dmaOp.getTargetObjectFifo()) &&
+                 sourceMemspace > targetMemspace) {
+        rewriter.moveOpAfter(dmaOp, forOp);
+        hoistHappened = true;
+        return WalkResult::advance();
+      }
+
+      // If this dma op can't be hoisted due to dependencies, keep adding new
+      // dependencies.
+      addDependencies(dmaOp);
+      return WalkResult::advance();
+    });
+    if (hoistHappened) return success();
+    return failure();
+  }
+
+  /// Unroll the for loop, skipping the first iteration. While unrolling, create
+  /// new clones of `LogicalObjectFifoFromMemrefOp` so they can be distributed
+  /// onto multiple physical locations later.
+  LogicalResult loopUnrollAndDistributeLogicalObjectFifos(
+      RewriterBase &rewriter, scf::ForOp forOp) const {
+    Block *loopBodyBlock = forOp.getBody();
+    OpBuilder builder = OpBuilder::atBlockTerminator(loopBodyBlock);
+
+    // Keep a pointer to the last non-terminator operation in the original block
+    // so that we know what to clone (since we are doing this in-place).
+    Block::iterator srcBlockEnd = std::prev(loopBodyBlock->end(), 2);
+
+    // Update loop bounds
+    int64_t lbInt = getConstantIntValue(forOp.getLowerBound()).value();
+    int64_t ubInt = getConstantIntValue(forOp.getUpperBound()).value();
+    int64_t stepInt = getConstantIntValue(forOp.getStep()).value();
+    if (lbInt != 0 || stepInt != 1) return failure();
+    if (stepInt == (ubInt - lbInt)) return failure();
```

Comment:
It's a good practice to put braces around trivial bodied `if`, `for` and `while`.

Reply:
I don't see this rule in the mlir/llvm codebase? I see `if/else` without braces everywhere.

Reply:
we follow google coding standard which iree and hence iree-amd-aie follows. If this was in core then we follow llvm standards which say no trivial braces all the time I think.
I think its a bit nuanced then just "always trivial braces". I think the way it is now is correct in this case becuase it fits on single line and there is no else part.
https://google.github.io/styleguide/cppguide.html#Formatting_Looping_Branching